indexbs/indexbacked_bs.go

```diff
@@ -0,0 +1,194 @@
+		val, ok := ro.blockstoreCache.Get(sk)
+		if !ok {
+			continue
+		}
+
+		rbs := val.(*accessorWithBlockstore).bs
+		blk, err := rbs.Get(ctx, c)
+		if err != nil {
+			// we know that the cid we want to lookup belongs to a shard with key `sk` and
+			// so if we fail to get the corresponding block from the blockstore for that shards, something has gone wrong
+			// and we should remove the blockstore for that shard from our cache.
+			ro.blockstoreCache.Remove(sk)
+			continue
+		}
+
+		// add the block to the block cache
+		logbs.Debugw("Get: returning from block store cache", "cid", c)
+		return blk, nil
+	}
+
+	// ---- we don't have a cached blockstore for a shard that can serve the block -> let's build one.
+
+	// select a valid shard that can serve the retrieval
+	sk, err := ro.shardSelectF(c, shards)
+	if err != nil {
+		return nil, ErrBlockNotFound
+	}
+
+	// load blockstore for the selected shard and try to serve the cid from that blockstore.
+	resch := make(chan dagstore.ShardResult, 1)
+	if err := ro.d.AcquireShard(ctx, sk, resch, dagstore.AcquireOpts{}); err != nil {
+		return nil, fmt.Errorf("failed to acquire shard %s: %w", sk, err)
+	}
+	var res dagstore.ShardResult
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case res = <-resch:
+		if res.Error != nil {
+			return nil, fmt.Errorf("failed to acquire shard %s: %w", sk, res.Error)
+		}
+	}
+
+	sa := res.Accessor
+	bs, err := sa.Blockstore()
+	if err != nil {
+		return nil, fmt.Errorf("failed to load read only blockstore for shard %s: %w", sk, err)
+	}
+
+	blk, err := bs.Get(ctx, c)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get block: %w", err)
+	}
+
+	// update the block cache and the blockstore cache
+	ro.blockstoreCache.Add(sk, &accessorWithBlockstore{sa, bs})
+
+	logbs.Debugw("Get: returning after creating new blockstore", "cid", c)
+	return blk, nil
+}
+
+func (ro *IndexBackedBlockstore) Has(ctx context.Context, c cid.Cid) (bool, error) {
+	logbs.Debugw("Has called", "cid", c)
+
+	// if there is a shard that can serve the retrieval for the given cid, we have the requested cid
+	// and has should return true.
+	shards, err := ro.d.ShardsContainingMultihash(ctx, c.Hash())
+	if err != nil {
+		logbs.Debugw("Has error", "cid", c, "err", err)
+		return false, nil
+	}
+	if len(shards) == 0 {
+		logbs.Debugw("Has: returning false no error", "cid", c)
+		return false, nil
+	}
+
+	_, err = ro.shardSelectF(c, shards)
+	if err != nil {
+		logbs.Debugw("Has error", "cid", c, "err", err)
+		return false, ErrBlockNotFound
+	}
+
+	logbs.Debugw("Has: returning true", "cid", c)
+	return true, nil
+}
+
+func (ro *IndexBackedBlockstore) GetSize(ctx context.Context, c cid.Cid) (int, error) {
+	logbs.Debugw("GetSize called", "cid", c)
+
+	blk, err := ro.Get(ctx, c)
+	if err != nil {
+		logbs.Debugw("GetSize error", "cid", c, "err", err)
+		return 0, fmt.Errorf("failed to get block: %w", err)
+	}
+
+	logbs.Debugw("GetSize success", "cid", c)
+	return len(blk.RawData()), nil
```

Comment:
Is it possible to call `GetSize()` on the underlying blockstore instead of loading the block's raw data into memory?
Bitswap uses the `GetSize()` call [on the hot path](https://github.com/ipfs/go-bitswap/blob/b6f0cc7c83aaa27a39cc7e1b16ee34bba2d8b5b8/internal/decision/engine.go#L622) so it's important that it be as efficient as possible.

Reply:
What do you mean by calling `GetSize()` on the  underlying blockstore?